Adding changelog for node compat improvements (#19341)
cloudflare/cloudflare-docs: src/content/changelogs-next/2025-01-22-nodejs-compat-improvements.mdx

<TypeScriptExample filename="index.ts">
```ts
import dns from 'node:dns';

dns.lookup('example.org', (err, address, ipFamily) => {
	console.log('address: %j family: IPv%s', address, ipFamily);
});
```
</TypeScriptExample>

All `node:dns` functions are available.

#### node:timers

You can use [`node:timers`](https://nodejs.org/api/timers.html) to schedule functions to be called at some future period of time.

This includes [`setTimeout`](https://nodejs.org/api/timers.html#settimeoutcallback-delay-args) for calling a function after a delay,
[`setInterval`](https://nodejs.org/api/timers.html#clearintervaltimeout) for calling a function repeatedly,
and [`setImmediate`](https://nodejs.org/api/timers.html#setimmediatecallback-args) for calling a function in the next iteration of the event loop.

<TypeScriptExample filename="index.ts">
```ts
import timers from 'node:timers';

console.log("first");
timers.setTimeout(() => {
	console.log("last");
}, 10);

timers.setTimeout(() => {
	console.log("next");
});
```

Note that due to [security-based restrictions on timers](/workers/reference/security-model/#step-1-disallow-timers-and-multi-threading) in Workers,

> timers are limited to returning the time of the last I/O. This means that while setTimeout and setInterval will defer your function execution

Reviewer: ```suggestion
timers are limited to returning the time of the last I/O. This means that while setTimeout, setInterval and setImmediate will defer your function execution
```

---

> title: Improved Node.js compatibility

Reviewer 2: Think title can get more specific

> description: Node.js APIs from the node:dns, node:net, and node:timers modules are now available when using nodejs_compatibility.

Reviewer 2: ```suggestion
description: Node.js APIs from the node:dns, node:net, and node:timers modules are now available when using nodejs_compat.
```

products:

>   - workers

Reviewer 3: also pages?

date: 2025-01-22T13:00:00Z
---

import { Render, PackageManagers, TypeScriptExample } from "~/components";

> When using a Worker with [nodejs_compatability](/workers/runtime-apis/nodejs/) enabled, you can now use the following Node.js APIs:

Reviewer 2: ```suggestion
When using a Worker with the [`nodejs_compat`](/workers/runtime-apis/nodejs/) compatibility flag enabled, you can now use the following Node.js APIs:
```

- [`node:net`](/workers/runtime-apis/nodejs/net/)
- [`node:dns`](/workers/runtime-apis/nodejs/dns/)
- [`node:timers`](/workers/runtime-apis/nodejs/timers/)

#### node:net

You can use [`net.Socket`](https://nodejs.org/api/net.html#class-netsocket) to create a direct connection to a server via a TCP socket.
This can be useful when connecting to databases such as [MySQL](https://www.mysql.com/) or [PostgreSQL](https://www.postgresql.org/),

> though if possible, using [Hyperdrive Connection Pooling](/hyperdrive/configuration/how-hyperdrive-works/#connection-pooling) is preferable.

Reviewer 2: This doesn't quite make sense to me since you need to connect to a DB securely using `tls`?

And we wouldn't ever actually recommend that someone connect to a DB from a Worker without Hyperdrive?

> These functions use [`connect`](/workers/runtime-apis/tcp-sockets/#connect) functionality from the built-in `cloudflare:sockets` module.

Reviewer 2: Why is this relevant to the reader? Is it because the use of `connect()` under the hood changes something about the APIs?

Author: Yeah good point - I might keep it in the docs so they know about that as a non-Node option, but not here

<TypeScriptExample filename="index.ts">
```ts
import net from 'node:net';

const exampleIP = '127.0.0.1';

export default {
	async fetch(req): Promise<Response> {
		const socket = new net.Socket();
		socket.connect(4000, exampleIP, function () {
			console.log('Connected');
		});

>     	socket.write('Hello, Server!');

Reviewer 3: nit: formatting looks off - `pnpm format` should help

    	socket.end();

    	return new Response('Wrote to server', { status: 200 });
    },

} satisfies ExportedHandler;

````
</TypeScriptExample>

Additionally, you can now use other APIs incliding [`net.BlockList`](https://nodejs.org/api/net.html#class-netblocklist) and
[`net.SocketAddress`](https://nodejs.org/api/net.html#class-netsocketaddress).

> Note that the [`net.Server`](https://nodejs.org/api/net.html#class-netserver) module is not supported.

Reviewer 3: I don't think this is a "module"

```suggestion
Note that [`net.Server`](https://nodejs.org/api/net.html#class-netserver) is not supported.
```

#### node:dns

> You can use [`node:dns`](https://nodejs.org/api/dns.html) for name resolution via DNS lookups using

Reviewer 2: Do you want to explain somewhere that this is DoH?

[Cloudflare DNS](https://www.cloudflare.com/application-services/products/dns/) at 1.1.1.1.

<TypeScriptExample filename="index.ts">
```ts
import dns from "node:dns";

> dns.lookup("example.org", (_err: any, address: string, ipFamily: number) => {
> 	console.log("address: %j family: IPv%s", address, ipFamily);

Reviewer 3: ```suggestion
dns.lookup("example.org", (_err: any, address: string, ipFamily: number) => 
	console.log(`address: ${address} family: IPv${ipFamily}`);
```

````

</TypeScriptExample>

> All `node:dns` functions are available.

Reviewer 3: not sure this is correct, @anonrig?

Reviewer 3: Yep, I think we should clearly indicate that those functions are throwing for now

#### node:timers

You can use [`node:timers`](https://nodejs.org/api/timers.html) to schedule functions to be called at some future period of time.

> This includes [`setTimeout`](https://nodejs.org/api/timers.html#settimeoutcallback-delay-args) for calling a function after a delay,

Reviewer 3: The links doesn't look they match the functions

Author: One was wrong (fixed now), but the others are good - the arguments in the function name get added to the URL, so it looks off

[`setInterval`](https://nodejs.org/api/timers.html#clearintervaltimeout) for calling a function repeatedly,
and [`setImmediate`](https://nodejs.org/api/timers.html#setimmediatecallback-args) for calling a function in the next iteration of the event loop.

<TypeScriptExample filename="index.ts">
```ts
import timers from 'node:timers';

console.log("first");
timers.setTimeout(() => {
console.log("last");
}, 10);

timers.setTimeout(() => {
console.log("next");
});

```
</TypeScriptExample>

Note that due to [security-based restrictions on timers](/workers/reference/security-model/#step-1-disallow-timers-and-multi-threading) in Workers,
timers are limited to returning the time of the last I/O. This means that while setTimeout, setInterval, and setImmediate will defer your function execution

> until after other events have run, they will not delay them for the full time specified.

Reviewer 2: This seems like more of a detail for the docs page about the API than the changelog post

Author: +1 - keeping in docs but removing from here
